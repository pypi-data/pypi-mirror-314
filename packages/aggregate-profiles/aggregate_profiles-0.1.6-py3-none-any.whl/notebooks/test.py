import aggregate_profiles as aggp 

print(dir(aggp))