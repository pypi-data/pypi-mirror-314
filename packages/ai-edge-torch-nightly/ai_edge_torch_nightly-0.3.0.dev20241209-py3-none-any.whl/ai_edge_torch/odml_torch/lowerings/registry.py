# Copyright 2024 The AI Edge Torch Authors.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ==============================================================================
"""Torch op decompositions and MLIR lowerings registry."""

from typing import Any, Callable

import torch

from . import context


class LoweringRegistry:
  """Registry object for torch op decompositions and to-MLIR lowerings."""

  def __init__(self):
    self.registered_ops = {}
    self.decompositions = {}

  def lookup(self, op_or_name):
    candidate = self._get_lowering(op_or_name)
    if candidate is None:
      if isinstance(op_or_name, torch._ops.OpOverloadPacket):
        candidate = self._get_lowering(op_or_name.default)
      if isinstance(op_or_name, torch._ops.OpOverload):
        candidate = self._get_lowering(op_or_name.overloadpacket)
    return candidate

  def _get_lowering(self, op):
    candidate = self.registered_ops.get(op)
    return candidate

  def register(self, op, lowering):
    if isinstance(op, torch._ops.OpOverloadPacket):
      ops = [getattr(op, overload) for overload in op.overloads()]
    else:
      ops = [op]

    for op in ops:
      self.registered_ops[op] = lowering


global_registry = LoweringRegistry()
global_registry.decompositions.update(torch._decomp.core_aten_decompositions())
global_registry.decompositions.update(
    torch._decomp.get_decompositions([
        torch.ops.aten.upsample_nearest2d,
        torch.ops.aten._native_batch_norm_legit.no_stats,
        torch.ops.aten._native_batch_norm_legit_functional,
        torch.ops.aten._adaptive_avg_pool2d,
        torch.ops.aten._adaptive_avg_pool3d,
        torch.ops.aten.grid_sampler_2d,
        torch.ops.aten.native_group_norm,
        torch.ops.aten.native_dropout,
        torch.ops.aten.reflection_pad1d,
        torch.ops.aten.reflection_pad2d,
        torch.ops.aten.reflection_pad3d,
        torch.ops.aten.replication_pad1d,
        torch.ops.aten.replication_pad2d,
        torch.ops.aten.replication_pad3d,
        torch.ops.aten.addmm,
    ])
)

torch._decomp.remove_decompositions(
    global_registry.decompositions,
    [
        torch.ops.aten.roll,
    ],
)


def lookup(op):
  return global_registry.lookup(op)


def lower(op):
  def inner(lowering: Callable[[context.LoweringContext, ...], Any]):
    global_registry.register(op, lowering)
    return lowering

  return inner


def decompositions():
  return global_registry.decompositions
