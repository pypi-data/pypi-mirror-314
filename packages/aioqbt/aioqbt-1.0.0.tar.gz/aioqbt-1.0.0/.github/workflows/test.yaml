---
name: test

on:
  push:

jobs:
  test:
    runs-on: ubuntu-latest
    strategy:
      fail-fast: false
      matrix:
        python-version: ["3.7", "3.8", "3.9", "3.10", "3.11", "3.12", "3.13"]
    steps:
      - name: Checkout
        uses: actions/checkout@v3

      - name: Set up Python ${{ matrix.python-version }}
        uses: actions/setup-python@v4
        with:
          python-version: ${{ matrix.python-version }}

      - name: Install qbittorrent-nox
        run: |
          sudo apt-get install --yes --no-install-recommends qbittorrent-nox

      - name: Install package
        run: |
          python -m pip install --upgrade pip
          python -m pip install .[test]

      - name: Run tests
        run: pytest
