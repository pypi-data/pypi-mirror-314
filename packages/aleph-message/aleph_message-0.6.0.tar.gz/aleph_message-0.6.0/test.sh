#!/bin/sh

set -euf

podman build --loglevel 3 -t aleph-message .
podman run --rm -ti -v $(pwd)/aleph_message:/opt/aleph_message aleph-message pytest -vv
podman run --rm -ti -v $(pwd)/aleph_message:/opt/aleph_message aleph-message mypy --ignore-missing-imports aleph_message
