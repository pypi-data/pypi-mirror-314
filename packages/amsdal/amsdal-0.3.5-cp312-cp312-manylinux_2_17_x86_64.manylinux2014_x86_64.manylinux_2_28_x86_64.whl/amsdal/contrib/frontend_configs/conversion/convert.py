from datetime import date
from datetime import datetime
from enum import Enum
from inspect import _empty
from inspect import signature
from types import FunctionType
from types import UnionType
from typing import Any
from typing import ClassVar
from typing import ForwardRef
from typing import Union

from amsdal_models.classes.manager import ClassManager
from amsdal_models.classes.model import LegacyModel
from amsdal_models.classes.model import Model
from amsdal_utils.models.data_models.reference import Reference
from pydantic import BaseModel
from pydantic_core import PydanticUndefined

from amsdal.schemas.manager import SchemaManager

default_types_map = {
    int: 'number',
    float: 'number',
    bool: 'checkbox',
    str: 'text',
    bytes: 'Bytes',
    date: 'date',
    datetime: 'datetime',
}


def _process_union(value: UnionType, *, is_transaction: bool = False) -> dict[str, Any]:
    arg_type = {'required': True}
    for arg in value.__args__:
        if arg is type(None):
            arg_type['required'] = False
            continue

        control = convert_to_frontend_config(arg, is_transaction=is_transaction)
        if control:
            arg_type.update(control)

    return arg_type


def convert_to_frontend_config(value: Any, *, is_transaction: bool = False) -> dict[str, Any]:
    """
    Converts a given value to a frontend configuration dictionary.

    This function takes a value and converts it into a dictionary that represents
    the configuration for a frontend form control. It handles various types such as
    Union, list, dict, BaseModel, and custom types.

    Args:
        value (Any): The value to be converted to frontend configuration.
        is_transaction (bool, optional): Indicates if the conversion is for a transaction. Defaults to False.

    Returns:
        dict[str, Any]: A dictionary representing the frontend configuration for the given value.
    """
    if hasattr(value, '__origin__'):
        origin_class = value.__origin__

        if origin_class in [ClassVar]:
            return {}

        if origin_class is Union:
            _union = _process_union(value, is_transaction=is_transaction)

            if 'entityType' in _union and _union['entityType'] == 'File':
                _union['type'] = 'file'
                del _union['entityType']

            return _union

        if origin_class is list:
            return {
                'type': 'array',
                'name': 'array_items',
                'label': 'array_items',
                'control': {
                    'name': 'array_items_values',
                    'label': 'array_items_values',
                    **convert_to_frontend_config(value.__args__[0], is_transaction=is_transaction),
                },
            }

        if origin_class is dict:
            return {
                'type': 'dict',
                'name': 'dict_items',
                'label': 'dict_items',
                'control': {
                    'name': 'dict_items_values',
                    'label': 'dict_items_values',
                    **convert_to_frontend_config(value.__args__[1], is_transaction=is_transaction),
                },
            }

    if isinstance(value, ForwardRef):
        class_name = value.__forward_arg__
        _class = ClassManager().import_class(class_name, ClassManager().resolve_schema_type(class_name))

        if issubclass(_class, Model):
            return {
                'entityType': value.__forward_arg__,
            }

        value = _class

    if isinstance(value, UnionType):
        _union = _process_union(value, is_transaction=is_transaction)

        if 'entityType' in _union and _union['entityType'] == 'File':
            _union['type'] = 'file'
            del _union['entityType']

        return _union

    if value in default_types_map:
        return {
            'type': default_types_map[value],
        }

    if value is Any:
        return {
            'type': 'text',
        }

    if isinstance(value, FunctionType):
        function_controls = []

        while hasattr(value, '__wrapped__'):
            value = value.__wrapped__

        _signature = signature(value)
        _parameters = _signature.parameters

        for arg_name in _parameters:
            if arg_name in value.__annotations__:
                arg_type = value.__annotations__[arg_name]
                control = convert_to_frontend_config(
                    arg_type,
                    is_transaction=True,
                )

                if not control:
                    continue

                control['name'] = arg_name
                control['label'] = arg_name
            else:
                control = {
                    'type': 'text',
                    'name': arg_name,
                    'label': arg_name,
                }

            control.setdefault('required', True)
            _param = _parameters[arg_name]
            if _param.default is not _empty:
                control['value'] = _param.default
                control['required'] = False

            function_controls.append(control)

        return {
            'type': 'group',
            'name': value.__name__,
            'label': value.__name__,
            'controls': function_controls,
        }

    if issubclass(value, Reference):
        return {
            'type': 'object_latest',
        }

    if is_transaction and issubclass(value, Model):
        if value.__name__ == 'File':
            return {
                'type': 'file',
            }
        return {
            'type': 'object_latest',
            'entityType': value.__name__,
        }

    if issubclass(value, LegacyModel):
        return {}

    if issubclass(value, BaseModel):
        model_controls = []

        try:
            schema = SchemaManager().get_schema_by_name(value.__name__)
        except FileNotFoundError:
            schema = None

        for field_name, field in value.__annotations__.items():
            control = convert_to_frontend_config(field, is_transaction=is_transaction)

            if not control:
                continue

            control.setdefault('required', True)

            if field_name in value.model_fields:
                _field = value.model_fields[field_name]

                if _field.default is not PydanticUndefined:
                    control['value'] = _field.default

            control['name'] = field_name
            control['label'] = field_name

            if schema and schema.properties and field_name in schema.properties:
                schema_property = schema.properties[field_name]

                if schema_property.options:
                    control['options'] = [
                        {
                            'label': option.key,
                            'value': option.value,
                        }
                        for option in schema_property.options
                    ]
                    if control.get('type') == 'text':
                        control['type'] = 'select'

                if schema_property.title:
                    control['label'] = schema_property.title

            model_controls.append(control)

        return {
            'type': 'group',
            'name': value.__name__,
            'label': value.__name__,
            'controls': model_controls,
        }

    if issubclass(value, Enum):
        return {
            'type': 'select',
            'options': [{'label': option.name, 'value': option.value} for option in value],
        }

    return {}
