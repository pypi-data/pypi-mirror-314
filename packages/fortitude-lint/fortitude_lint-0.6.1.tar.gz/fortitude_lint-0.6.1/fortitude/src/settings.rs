use ruff_diagnostics::Applicability;
use ruff_macros::CacheKey;
use serde::{Deserialize, Serialize};
/// A collection of user-modifiable settings. Should be expanded as new features are added.
use std::fmt::{Display, Formatter};

use crate::rule_selector::RuleSelector;

pub struct Settings {
    pub line_length: usize,
}

impl Default for Settings {
    fn default() -> Self {
        Self { line_length: 100 }
    }
}

/// Toggle for rules still in preview
#[derive(Debug, Clone, Copy, PartialEq, Eq, Default, CacheKey, is_macro::Is)]
pub enum PreviewMode {
    #[default]
    Disabled,
    Enabled,
}

impl From<bool> for PreviewMode {
    fn from(version: bool) -> Self {
        if version {
            PreviewMode::Enabled
        } else {
            PreviewMode::Disabled
        }
    }
}

impl Display for PreviewMode {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Disabled => write!(f, "disabled"),
            Self::Enabled => write!(f, "enabled"),
        }
    }
}

/// Default rule selection
pub const DEFAULT_SELECTORS: &[RuleSelector] = &[RuleSelector::All];

/// Toggle for unsafe fixes.
/// `Hint` will not apply unsafe fixes but a message will be shown when they are available.
/// `Disabled` will not apply unsafe fixes or show a message.
/// `Enabled` will apply unsafe fixes.
#[derive(Debug, Copy, Clone, CacheKey, Default, PartialEq, Eq, is_macro::Is)]
pub enum UnsafeFixes {
    #[default]
    Hint,
    Disabled,
    Enabled,
}

impl Display for UnsafeFixes {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::Hint => "hint",
                Self::Disabled => "disabled",
                Self::Enabled => "enabled",
            }
        )
    }
}

impl From<bool> for UnsafeFixes {
    fn from(value: bool) -> Self {
        if value {
            UnsafeFixes::Enabled
        } else {
            UnsafeFixes::Disabled
        }
    }
}

impl UnsafeFixes {
    pub fn required_applicability(&self) -> Applicability {
        match self {
            Self::Enabled => Applicability::Unsafe,
            Self::Disabled | Self::Hint => Applicability::Safe,
        }
    }
}

/// Toggle for progress bar
#[derive(
    Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Debug, Hash, Default, clap::ValueEnum,
)]
#[serde(rename_all = "kebab-case")]
pub enum ProgressBar {
    #[default]
    Off,
    Fancy,
    Ascii,
}

impl Display for ProgressBar {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        write!(
            f,
            "{}",
            match self {
                Self::Off => "off",
                Self::Fancy => "fancy",
                Self::Ascii => "ascii",
            }
        )
    }
}

#[derive(
    Clone, Copy, PartialEq, Eq, Serialize, Deserialize, Debug, Hash, Default, clap::ValueEnum,
)]
#[serde(rename_all = "kebab-case")]
pub enum OutputFormat {
    Concise,
    #[default]
    Full,
    Json,
    JsonLines,
    Junit,
    Grouped,
    Github,
    Gitlab,
    Pylint,
    Rdjson,
    Azure,
    Sarif,
}

impl Display for OutputFormat {
    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
        match self {
            Self::Concise => write!(f, "concise"),
            Self::Full => write!(f, "full"),
            Self::Json => write!(f, "json"),
            Self::JsonLines => write!(f, "json_lines"),
            Self::Junit => write!(f, "junit"),
            Self::Grouped => write!(f, "grouped"),
            Self::Github => write!(f, "github"),
            Self::Gitlab => write!(f, "gitlab"),
            Self::Pylint => write!(f, "pylint"),
            Self::Rdjson => write!(f, "rdjson"),
            Self::Azure => write!(f, "azure"),
            Self::Sarif => write!(f, "sarif"),
        }
    }
}

#[derive(Debug, Copy, Clone, Hash, is_macro::Is)]
pub enum FixMode {
    Generate,
    Apply,
    #[allow(dead_code)]
    Diff,
}
