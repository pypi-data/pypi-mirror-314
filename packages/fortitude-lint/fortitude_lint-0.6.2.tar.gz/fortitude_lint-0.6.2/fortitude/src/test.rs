use std::path::Path;

use anyhow::Result;
use ruff_source_file::{SourceFile, SourceFileBuilder};

use crate::{
    check::{
        ast_entrypoint_map, check_file, read_to_string, rules_to_path_rules, rules_to_text_rules,
    },
    message::{Emitter, TextEmitter},
    rules::Rule,
    settings::{FixMode, Settings, UnsafeFixes},
};

pub(crate) fn test_resource_path(path: impl AsRef<Path>) -> std::path::PathBuf {
    Path::new("./resources/test/").join(path)
}

/// Run [`check_path`] on a Fortran file in the `resources/test/fixtures` directory.
pub(crate) fn test_path(
    path: impl AsRef<Path>,
    rules: &[Rule],
    settings: &Settings,
) -> Result<String> {
    let path = test_resource_path("fixtures").join(path);
    let source = read_to_string(&path)?;
    let filename = path.to_string_lossy();
    let file = SourceFileBuilder::new(filename.as_ref(), source.as_str()).finish();
    Ok(test_contents(&path, &file, rules, settings))
}

pub(crate) fn test_contents(
    path: &Path,
    file: &SourceFile,
    rules: &[Rule],
    settings: &Settings,
) -> String {
    let path_rules = rules_to_path_rules(rules);
    let text_rules = rules_to_text_rules(rules);
    let ast_entrypoints = ast_entrypoint_map(rules);

    match check_file(
        &path_rules,
        &text_rules,
        &ast_entrypoints,
        path,
        file,
        settings,
        FixMode::Generate,
        UnsafeFixes::Hint,
    ) {
        Ok(violations) => {
            if violations.messages.is_empty() {
                return String::new();
            }

            let mut output = Vec::new();

            TextEmitter::default()
                .with_show_fix_status(true)
                .with_show_fix_diff(true)
                .with_show_source(true)
                .with_unsafe_fixes(crate::settings::UnsafeFixes::Enabled)
                .emit(&mut output, &violations.messages)
                .unwrap();

            String::from_utf8(output).unwrap()
        }
        Err(msg) => {
            panic!("Failed to process: {msg}");
        }
    }
}
