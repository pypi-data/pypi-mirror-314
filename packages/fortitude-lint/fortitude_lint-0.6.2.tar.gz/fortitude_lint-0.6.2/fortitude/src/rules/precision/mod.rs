pub mod double_precision;
pub mod implicit_kinds;
pub mod kind_suffixes;

#[cfg(test)]
mod tests {
    use std::convert::AsRef;
    use std::path::Path;

    use anyhow::Result;
    use insta::assert_snapshot;
    use test_case::test_case;

    use crate::registry::Rule;
    use crate::settings::Settings;
    use crate::test::test_path;

    #[test_case(Rule::NoRealSuffix, Path::new("P001.f90"))]
    #[test_case(Rule::DoublePrecision, Path::new("P011.f90"))]
    #[test_case(Rule::ImplicitRealKind, Path::new("P021.f90"))]
    fn rules(rule_code: Rule, path: &Path) -> Result<()> {
        let snapshot = format!("{}_{}", rule_code.as_ref(), path.to_string_lossy());
        let diagnostics = test_path(
            Path::new("precision").join(path).as_path(),
            &[rule_code],
            &Settings::default(),
        )?;
        assert_snapshot!(snapshot, diagnostics);
        Ok(())
    }
}
