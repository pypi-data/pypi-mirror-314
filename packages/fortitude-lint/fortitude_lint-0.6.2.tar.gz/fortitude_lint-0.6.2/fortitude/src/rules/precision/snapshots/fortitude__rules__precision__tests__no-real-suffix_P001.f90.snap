---
source: fortitude/src/rules/precision/mod.rs
expression: diagnostics
snapshot_kind: text
---
./resources/test/fixtures/precision/P001.f90:13:30: P001 real literal 1.23456 missing kind suffix
   |
11 |   real(sp), parameter :: e = 1.0e10 ! Okay: No loss of precision, e exponent
12 |   real(sp), parameter :: f = -2E10 ! Okay: No loss of precision, E exponent
13 |   real(sp), parameter :: g = 1.23456 ! Bad: Loss of precision
   |                              ^^^^^^^ P001
14 |   real(sp), parameter :: h = 1.23456e1 ! Bad: Loss of precision, e exponent
15 |   real(sp), parameter :: i = 1.23456E1 ! Bad: Loss of precision, E exponent
   |

./resources/test/fixtures/precision/P001.f90:14:30: P001 real literal 1.23456e1 missing kind suffix
   |
12 |   real(sp), parameter :: f = -2E10 ! Okay: No loss of precision, E exponent
13 |   real(sp), parameter :: g = 1.23456 ! Bad: Loss of precision
14 |   real(sp), parameter :: h = 1.23456e1 ! Bad: Loss of precision, e exponent
   |                              ^^^^^^^^^ P001
15 |   real(sp), parameter :: i = 1.23456E1 ! Bad: Loss of precision, E exponent
16 |   real(dp), parameter :: j = -1.23456_dp ! Okay: Kind suffix
   |

./resources/test/fixtures/precision/P001.f90:15:30: P001 real literal 1.23456E1 missing kind suffix
   |
13 |   real(sp), parameter :: g = 1.23456 ! Bad: Loss of precision
14 |   real(sp), parameter :: h = 1.23456e1 ! Bad: Loss of precision, e exponent
15 |   real(sp), parameter :: i = 1.23456E1 ! Bad: Loss of precision, E exponent
   |                              ^^^^^^^^^ P001
16 |   real(dp), parameter :: j = -1.23456_dp ! Okay: Kind suffix
17 |   real(sp), parameter :: k = 1.23456_sp ! Okay: Loss of precision, but we're explicit
   |

./resources/test/fixtures/precision/P001.f90:20:30: P001 real literal 2e39 missing kind suffix
   |
18 |   real(dp), parameter :: l = 1.23456d1 ! Okay: Ignore d exponent
19 |   real(dp), parameter :: m = 1.23456D3 ! Okay: Ignore D exponent
20 |   real(dp), parameter :: n = 2e39 ! Bad: Magnitude doesn't fit in single precision
   |                              ^^^^ P001
21 |   real(dp), parameter :: o = -(+(-(((3.141592654))))) ! Bad: Loss of precision, horrible declaration
   |

./resources/test/fixtures/precision/P001.f90:21:38: P001 real literal 3.141592654 missing kind suffix
   |
19 |   real(dp), parameter :: m = 1.23456D3 ! Okay: Ignore D exponent
20 |   real(dp), parameter :: n = 2e39 ! Bad: Magnitude doesn't fit in single precision
21 |   real(dp), parameter :: o = -(+(-(((3.141592654))))) ! Bad: Loss of precision, horrible declaration
   |                                      ^^^^^^^^^^^ P001
22 |   
23 |   complex(dp), parameter :: ca = (0.0, 0.0) ! Okay: No loss of precision
   |

./resources/test/fixtures/precision/P001.f90:29:35: P001 real literal 1.23456 missing kind suffix
   |
27 |   complex(sp), parameter :: ce = (1.0e10, 2.0e5) ! Okay: No loss of precision, e exponent
28 |   complex(sp), parameter :: cf = (2E10, -4.0e-5) ! Okay: No loss of precision, E exponent
29 |   complex(sp), parameter :: cg = (1.23456, -0.0) ! Bad: Loss of precision
   |                                   ^^^^^^^ P001
30 |   complex(sp), parameter :: ch = (0.0, 1.23456e1) ! Bad: Loss of precision, e exponent
31 |   complex(sp), parameter :: ci = (1.23456E1, 0.0) ! Bad: Loss of precision, E exponent
   |

./resources/test/fixtures/precision/P001.f90:30:40: P001 real literal 1.23456e1 missing kind suffix
   |
28 |   complex(sp), parameter :: cf = (2E10, -4.0e-5) ! Okay: No loss of precision, E exponent
29 |   complex(sp), parameter :: cg = (1.23456, -0.0) ! Bad: Loss of precision
30 |   complex(sp), parameter :: ch = (0.0, 1.23456e1) ! Bad: Loss of precision, e exponent
   |                                        ^^^^^^^^^ P001
31 |   complex(sp), parameter :: ci = (1.23456E1, 0.0) ! Bad: Loss of precision, E exponent
32 |   complex(dp), parameter :: cj = (-1.23456_dp, 0.2_dp) ! Okay: Kind suffix
   |

./resources/test/fixtures/precision/P001.f90:31:35: P001 real literal 1.23456E1 missing kind suffix
   |
29 |   complex(sp), parameter :: cg = (1.23456, -0.0) ! Bad: Loss of precision
30 |   complex(sp), parameter :: ch = (0.0, 1.23456e1) ! Bad: Loss of precision, e exponent
31 |   complex(sp), parameter :: ci = (1.23456E1, 0.0) ! Bad: Loss of precision, E exponent
   |                                   ^^^^^^^^^ P001
32 |   complex(dp), parameter :: cj = (-1.23456_dp, 0.2_dp) ! Okay: Kind suffix
33 |   complex(sp), parameter :: ck = (1.23456_sp, 3.14159265_sp)! Okay: Loss of precision, but we're explicit
   |

./resources/test/fixtures/precision/P001.f90:36:35: P001 real literal 2e39 missing kind suffix
   |
34 |   complex(dp), parameter :: cl = (1.23456d1, 37d2) ! Okay: Ignore d exponent
35 |   complex(dp), parameter :: cm = (1.23456D3, 37D2) ! Okay: Ignore D exponent
36 |   complex(dp), parameter :: cn = (2e39, 0e0) ! Bad: Magnitude doesn't fit in single precision
   |                                   ^^^^ P001
37 |   real(dp), parameter :: co = -(+(-(-3.141592654, +5.182647))) ! Bad: Loss of precision, horrible declaration
   |

./resources/test/fixtures/precision/P001.f90:37:38: P001 real literal 3.141592654 missing kind suffix
   |
35 |   complex(dp), parameter :: cm = (1.23456D3, 37D2) ! Okay: Ignore D exponent
36 |   complex(dp), parameter :: cn = (2e39, 0e0) ! Bad: Magnitude doesn't fit in single precision
37 |   real(dp), parameter :: co = -(+(-(-3.141592654, +5.182647))) ! Bad: Loss of precision, horrible declaration
   |                                      ^^^^^^^^^^^ P001
38 |
39 |   real(dp) :: p, q, x, y, z
   |

./resources/test/fixtures/precision/P001.f90:37:52: P001 real literal 5.182647 missing kind suffix
   |
35 |   complex(dp), parameter :: cm = (1.23456D3, 37D2) ! Okay: Ignore D exponent
36 |   complex(dp), parameter :: cn = (2e39, 0e0) ! Bad: Magnitude doesn't fit in single precision
37 |   real(dp), parameter :: co = -(+(-(-3.141592654, +5.182647))) ! Bad: Loss of precision, horrible declaration
   |                                                    ^^^^^^^^ P001
38 |
39 |   real(dp) :: p, q, x, y, z
   |

./resources/test/fixtures/precision/P001.f90:42:12: P001 real literal 2.0 missing kind suffix
   |
40 |   complex(dp) :: cp, cq, cx, cy, cz
41 |
42 |   x = sqrt(2.0) ! Bad: Loss of precision
   |            ^^^ P001
43 |   y = real(1.0, kind=dp) ! Okay: Type cast with no loss of precision
44 |   z = real(1.0 + 1.0, kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |

./resources/test/fixtures/precision/P001.f90:44:12: P001 real literal 1.0 missing kind suffix
   |
42 |   x = sqrt(2.0) ! Bad: Loss of precision
43 |   y = real(1.0, kind=dp) ! Okay: Type cast with no loss of precision
44 |   z = real(1.0 + 1.0, kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |            ^^^ P001
45 |   p = real(-5.0, kind=dp) ! Okay: Type cast with no loss of precision
46 |   q = real(1.23456, kind=dp) ! Bad: Type cast with loss of precision
   |

./resources/test/fixtures/precision/P001.f90:44:18: P001 real literal 1.0 missing kind suffix
   |
42 |   x = sqrt(2.0) ! Bad: Loss of precision
43 |   y = real(1.0, kind=dp) ! Okay: Type cast with no loss of precision
44 |   z = real(1.0 + 1.0, kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |                  ^^^ P001
45 |   p = real(-5.0, kind=dp) ! Okay: Type cast with no loss of precision
46 |   q = real(1.23456, kind=dp) ! Bad: Type cast with loss of precision
   |

./resources/test/fixtures/precision/P001.f90:46:12: P001 real literal 1.23456 missing kind suffix
   |
44 |   z = real(1.0 + 1.0, kind=dp) ! Bad: Type cast from expression, possible l.o.p
45 |   p = real(-5.0, kind=dp) ! Okay: Type cast with no loss of precision
46 |   q = real(1.23456, kind=dp) ! Bad: Type cast with loss of precision
   |            ^^^^^^^ P001
47 |   
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
   |

./resources/test/fixtures/precision/P001.f90:48:14: P001 real literal 2.0 missing kind suffix
   |
46 |   q = real(1.23456, kind=dp) ! Bad: Type cast with loss of precision
47 |   
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
   |              ^^^ P001
49 |   cy = cmplx((1.0, 1.0), kind=dp) ! Okay: Type cast with no loss of precision
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |

./resources/test/fixtures/precision/P001.f90:48:19: P001 real literal 0.0 missing kind suffix
   |
46 |   q = real(1.23456, kind=dp) ! Bad: Type cast with loss of precision
47 |   
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
   |                   ^^^ P001
49 |   cy = cmplx((1.0, 1.0), kind=dp) ! Okay: Type cast with no loss of precision
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |

./resources/test/fixtures/precision/P001.f90:50:15: P001 real literal 1.0 missing kind suffix
   |
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
49 |   cy = cmplx((1.0, 1.0), kind=dp) ! Okay: Type cast with no loss of precision
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |               ^^^ P001
51 |   cp = cmplx((5.0, -0.0625), kind=dp) ! Okay: Type cast with no loss of precision
52 |   cq = cmplx((-1.23456, 3.141292654), kind=dp) ! Bad: Type cast with loss of precision
   |

./resources/test/fixtures/precision/P001.f90:50:20: P001 real literal 0.0 missing kind suffix
   |
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
49 |   cy = cmplx((1.0, 1.0), kind=dp) ! Okay: Type cast with no loss of precision
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |                    ^^^ P001
51 |   cp = cmplx((5.0, -0.0625), kind=dp) ! Okay: Type cast with no loss of precision
52 |   cq = cmplx((-1.23456, 3.141292654), kind=dp) ! Bad: Type cast with loss of precision
   |

./resources/test/fixtures/precision/P001.f90:50:28: P001 real literal 0.0 missing kind suffix
   |
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
49 |   cy = cmplx((1.0, 1.0), kind=dp) ! Okay: Type cast with no loss of precision
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |                            ^^^ P001
51 |   cp = cmplx((5.0, -0.0625), kind=dp) ! Okay: Type cast with no loss of precision
52 |   cq = cmplx((-1.23456, 3.141292654), kind=dp) ! Bad: Type cast with loss of precision
   |

./resources/test/fixtures/precision/P001.f90:50:33: P001 real literal 1.0 missing kind suffix
   |
48 |   cx = sqrt((2.0, 0.0)) ! Bad: Loss of precision
49 |   cy = cmplx((1.0, 1.0), kind=dp) ! Okay: Type cast with no loss of precision
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
   |                                 ^^^ P001
51 |   cp = cmplx((5.0, -0.0625), kind=dp) ! Okay: Type cast with no loss of precision
52 |   cq = cmplx((-1.23456, 3.141292654), kind=dp) ! Bad: Type cast with loss of precision
   |

./resources/test/fixtures/precision/P001.f90:52:16: P001 real literal 1.23456 missing kind suffix
   |
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
51 |   cp = cmplx((5.0, -0.0625), kind=dp) ! Okay: Type cast with no loss of precision
52 |   cq = cmplx((-1.23456, 3.141292654), kind=dp) ! Bad: Type cast with loss of precision
   |                ^^^^^^^ P001
53 | end program test
   |

./resources/test/fixtures/precision/P001.f90:52:25: P001 real literal 3.141292654 missing kind suffix
   |
50 |   cz = cmplx((1.0, 0.0) + (0.0, 1.0), kind=dp) ! Bad: Type cast from expression, possible l.o.p
51 |   cp = cmplx((5.0, -0.0625), kind=dp) ! Okay: Type cast with no loss of precision
52 |   cq = cmplx((-1.23456, 3.141292654), kind=dp) ! Bad: Type cast with loss of precision
   |                         ^^^^^^^^^^^ P001
53 | end program test
   |
